helix-core/src/main/java/org/apache/helix/manager/zk/CallbackHandler.java

```diff
@@ -340,26 +340,29 @@ public void enqueueTask(NotificationContext changeContext) throws Exception {
   public void invoke(NotificationContext changeContext) throws Exception {
     Type type = changeContext.getType();
     long start = System.currentTimeMillis();
+    if (logger.isInfoEnabled()) {
+      logger.info("{} START: CallbackHandler {}, INVOKE {} listener: {} type: {}",
+          Thread.currentThread().getId(), _uid, _path, _listener, type);
+    }
 
-    // This allows the listener to work with one change at a time
-    synchronized (_manager) {
-      if (logger.isInfoEnabled()) {
-        logger
-            .info("{} START: CallbackHandler {}, INVOKE {} listener: {} type: {}", Thread.currentThread().getId(),
-                _uid, _path, _listener, type);
-      }
-
+    // TODO: Having subscribeForChanges in critical section might be overkill.
+    // Maybe it could be moved out later.
+    synchronized (this) {
```

Comment:
Why do we sync on this here? In fact, how do we reason it?

Assuming no batch, the calling path would be from enqueueTask(), which is called by handleXXX from ZkClient event thread. And also init() and reset()

Assuming batch, the calling path would be from _batchCallbackProcessor and also init() and reset().

init() is further called by either first time CallbackHandler construction or ZK session change.

It seems that ZK session change is the only places that there would be race condition, right?

Reply:
I think you already answered the question yourself.